tables/de-chardefs6.cti

```diff
@@ -1,6 +1,7 @@
-# liblouis: Character definitions for german tables
+# liblouis: Character definitions for german tables (bidirectional)
```

Comment:
Are we sure that the changes to de-chardefs6.cti don't affect the main table?

Reply:
> Yes. All tests pass, or did until we rebased.

Yes, I'm sure the tests pass, but the tests aren't that great. We only have a dictionary test. That means only words are tested. Anything special, like punctuation, is not covered.

I'm just looking at the file and while a lot of the changes are things that you just moved around, or where you added a `noback`, there also seem to be some actual changes in the dot patterns.

Reply:
> Ok, I actually just found something:
>
> With the original tables without any additional display table, accented letters now translate as ⠈letter, not “letter, i.e. ⠈a, not “a. This is an example of a dot pattern that needs to be mapped by an included display file, which I apparently forgot to do with the original tables.
>
> It actually should have been caught by the dictionary test, which contains words with accented letters. However, the yaml test file specifies its own display file and thereby masks the problem.
>
> I should include the display file in the original tables, same as in the bidi tables.

Nope. Wrong thinking :-) It is not a problem in the translation table. It is a problem in how you provide the braille input for a translation. The YAML test does it correctly: the braille encoding in the test matches the specified display table. You need to do the same.

Reply:
> Then, what solution would you suggest? Please, suggest something concrete instead of just saying “no”.

It's very simple. It's a matter of encoding/decoding braille correctly when interfacing with Liblouis. I'm not sure how you are invoking Liblouis, but if you either specify a display table and encode/decode your braille according to that table, or if you use the dotsIO mode, and you still get an unexpected result, _then_ there is a problem. Otherwise there is no problem.

Reply:
> Yes, I agree, and all is good and well, as long as the caller specifies a display table, which, unfortunately, they don’t at the moment. I know for certain that both the HIMS note-takers and NVDA call liblouis without specifying a display table, and I am almost sure about JAWS, even though I don’t have access to the source code. Personally, I test these things with lou_trace, with or without a display table.
>
> For a long term solution, I would suggest that the caller should be required to specify a display table, or that the table can specify a “default display table” to use with that table. No, this is not mixing the display opcode back with the others. It is simply a way for the table author to specify the default Braille representation for the table. This is typically guided by nationality and language. For Danish it would be da-dk-octobraille.dis. For German, it would typically be de-eurobrl6.dis.
>
> Yes, I am also looking forward to the day when we all use Unicode Braille for our embossers and notetakers, but, unfortunately, that is not the present situation.
>
> For a short term solution, I still suggest we include de-eurobrl6.dis in the German tables, even if it is “wrong thinking”. Then, they still work without specifying a display table explicitly, which is unfortunately required at the moment. If you have a better practical solution for now, please, tell me about it.

My solution is to force programs that use Liblouis to specify a display table. The reality is that if we do nothing, HIMS, NVDA, etc. are forever going to keep doing it like they do it today. It is just laziness from their part. There is no reason to wait for that. Sometimes you just have to make blunt decisions to get something going.

I don't even understand how a program like NVDA can get away with what they do (if it is indeed like you say)? Let's for a minute assume that you can be sure that everyone who uses the Danish Liblouis table has their braille device configured to use the character set that matches exactly octobraille.dis (which is hard to believe, but OK). What do you do if you want to read German text with NVDA and for that you switch to the German table? Do you change the settings of your braille display accordingly? That would be terrible!

But OK, you convinced me, we can include de-eurobrl6.dis in the German table. As long as it's allowed it is indeed better to not unnecessarily break the current behavior. Before we stop allowing using a translation tables as display table, we should make sure we replace or convert them with proper display tables anyway, and by including de-eurobrl6.dis we already know what the replace it with when the time comes.

By the way it is probably also a good idea to start add metadata to the display tables. It will become useful.

Reply:
> Luckily, the situation is not quite that bad when using a screen reader with a display. You don’t configure the character/display table on the display, but only on the computer or tablet/phone. I am not quite sure about the protocol for communication between a screen reader and a Braille display, but it seems to be communicating dots rather than their character equivalents. It is an old serial protocol, which probably hasn’t changed since the first Braille displays in the 90s.
>
> However, the situation with embossers and note-takers is more or less like you describe it. If I receive a .brl file from a German, I will have to change my note-taker or embosser to use the correct table before trying to read or emboss the file. And yes, all embossers in Denmark are set up to use the OctoBraille table, unless you come across a really old installation that was set up to use US computer Braille. This is why I am looking forward to all of us just using Unicode Braille and nothing else.

So if we're just talking about screen readers, they _must_ be specifying a display table or using the dotsIO mode, right? Otherwise they have no idea which dots they are sending to the braille display, and the protocol apparently requires it.

Even if the situation with embossers and note-takers is like it is, my standpoint is still that any software that makes use of Liblouis, if it can update to new versions of Liblouis, can also be changed to specify display tables.